File: CppPractice/Project1/Project1/Leetcode1512.cpp
#include <string>

#include <vector>

#include <iostream>

#include <stdlib.h>

#include <time.h>

#include <sstream>

#include <algorithm>

#include <map>

using namespace std;


// first solution is double for loop


class Solution {
public:
    int numIdenticalPairs(vector<int>& nums) {
        int ans=0;
        for (size_t i = 0; i < nums.size()-1; i++)
        {
            for (int j = i+1;  j< nums.size(); j++)
            {
                if (nums[j]==nums[i])
                {
                    ans += 1;
                }
            }
        }
        return ans;
    }
};


// second solution based on map


class Solution {
public:
    int numIdenticalPairs(vector<int>& nums) {
        map<int, int> intKey;
        int res = 0;
        for (int each : nums)
        {
            if (intKey.find(each) == intKey.end())
            {
                intKey[each] = 0;
            }
            intKey[each] += 1;
            res += (intKey[each] - 1);
        }
        return res;
    }

};

File: Pythonleetcode/Leetcode1512.py
class Solution(object):
    def numIdenticalPairs(self, nums):
        def conbinition_num(y):
            if y<2:
                return 0
            else:
                return int(y*(y-1)/2)

        nums=sorted(nums)
        ans=0
        for i in set(nums):
            ans += conbinition_num(nums.count(i))

        return ans



print(Solution().numIdenticalPairs([1,2,3,1,1,3]))
print(Solution().numIdenticalPairs([1,1,1,1]))
print(Solution().numIdenticalPairs([1,2,3]))